### PR TITLE
AWS: support node port health check

### PR DESCRIPTION
**Pull request by foolusion**
**What this PR does / why we need it**:
if a custom health check is set from the beta annotation on a service it
should be used for the ELB health check. This patch adds support for
that.
**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:
Let me know if any tests need to be added.
**Release note**:

```release-note
```

### CODE REVIEW

**Comment by jsafrane** [comment]
@k8s-bot ok to test

**Comment by jsafrane** [comment]
I don't know anything about load balancers and AWS health checks.
/assign @justinsb 
/unassign @jsafrane

**Comment by foolusion** [comment]
@justinsb can you please take a look

**Comment by foolusion** [comment]
@jsafrane @justinsb can I get an update on this?

**Comment by justinsb** [comment]
Can you describe what you're trying to archive @foolusion ?

I think this annotation was intended for use with ESIPP - so it's actually probably good to get it in.  But @thockin can you comment on whether we should be supporting this annotation separately from ESIPP, and who is looking after this?

I'll comment on a potential code refactoring directly on the code :-)

File: pkg/cloudprovider/providers/aws/aws.go [hunk 1]
```diff
@@ -2737,9 +2737,18 @@ func (c *Cloud) EnsureLoadBalancer(clusterName string, apiService *v1.Service, n
...
-	if err != nil {
-		return nil, err
+	if path, healthCheckNodePort := service.GetServiceHealthCheckPathPort(apiService); path != "" {
+		glog.V(4).Infof("service %v (%v) needs health checks on :%d%s)", apiService.Name, loadBalancerName, healthCheckNodePort, path)
+		err = c.ensureHttpHealthCheck(loadBalancer, path, healthCheckNodePort)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to ensure health check for localized service %v on node port %v: %v", loadBalancerName, healthCheckNodePort, err)
+		}
+	} else {
+		glog.V(4).Infof("service %v does not need health checks", apiService.Name)
```
Comment (justinsb) [inline pkg/cloudprovider/providers/aws/aws.go]: Probably better to say "does not need custom health checks" or "uses default health checks" or similar

File: pkg/cloudprovider/providers/aws/aws_loadbalancer.go [hunk 2]
```diff
@@ -377,6 +377,49 @@ func (c *Cloud) ensureLoadBalancerHealthCheck(loadBalancer *elb.LoadBalancerDesc
...
+
+	// Default AWS settings
+	expectedHealthyThreshold := int64(2)
+	expectedUnhealthyThreshold := int64(6)
+	expectedTimeout := int64(5)
+	expectedInterval := int64(10)
+
+	expectedTarget := "HTTP:" + strconv.FormatInt(int64(port), 10) + path
+
+	if expectedTarget == orEmpty(actual.Target) &&
```
Comment (justinsb) [inline pkg/cloudprovider/providers/aws/aws_loadbalancer.go]: AFAICT from this point on the two health-check configuration functions are the same (there's a lot of overlap earlier also!)  Can we combine the two tails here into a single function?  Maybe even just have one function, and build the expected healthCheck in EnsureLoadBalancer ?

**Comment by foolusion** [comment]
I'm currently using this to eliminate additional hops when using a load balancer in AWS. Currently the ELB only checks if it can make a connection to the service but this would allow more sophisticated health reporting.

**Comment by foolusion** [comment]
@justinsb I've combined those two methods and updated the log output. The main change is moving the logic for TCP health check into the aws.go EnsureLoadBalacner method and changing the function signature of ensureLoadBalancerHealthCheck. Please take another look.

**Comment by emmanuel** [comment]
We (@Nordstrom/Kubernetes) are very interested in this feature. Is there anything we can do to help get this merged?

**Comment by thockin** [comment]
@mrhohn @nicksardo

**Comment by emmanuel** [comment]
I should add that I (used to) work with @foolusion and that I believe he won't be available to speak up for this PR any further (he's off on other, wonderful, adventures). So I'm roughly proposing that I, @dhawal55, @sleepybrett and @mariusgrigoriu carry this PR forward in his stead (all coworkers of foolusion's, till recently).

**Comment by MrHohn** [comment]
@emmanuel This is great, I'd happy to see folks are extending [the OnlyLocal feature](https://github.com/kubernetes/features/issues/27) to AWS :)

**Comment by MrHohn** [comment]
To better clarify current status of ESIPP (External Source IP Preservation):
- ESIPP is going to GA in v1.7 - #41162.
- The beta annotations (`service.beta.kubernetes.io/external-traffic` and `service.beta.kubernetes.io/healthcheck-nodeport`) have been promoted as first class fields (`Service.Spec.ExternalTrafficPolicy` and `Service.Spec.HealthCheckNodePort`).
- For now this feature is only supported on GCE.

I'm not sure how would AWS LB behave when forwarding traffic onto nodes (preserving traffic or not). But only for preventing the second hop, I think having the health check properly set up on LB would be sufficient.

**Comment by foolusion** [comment]
I'm still partially available (when i have wifi), just in GMT timezone now. Let me know if you need me to make any further changes.

**Comment by thockin** [comment]
Just to be clear.  ESIPP is what Google called it, because that was the
customer demand.  It doesn't preserve IP for ELB because ELB never carried
the client IP in the first place (except in PROXY headers, which kube-proxy
does not deal with).

Only local is still useful, but it's not ESIPP.

And yes, we really do need it to work on other platforms.  Fragmentation
bad.

**Comment by emmanuel** [comment]
ESIPP sounds lovely and we'd be delighted to have it. That said, we're on AWS and (simply) eliminating the extra hop would be welcome. We're aquainted with `PROXY` protocol shenanigans for the ESIPP portion.

**Comment by justinsb** [comment]
Got it!  This sounds great, and now that it is a field my concerns about using the esipp annotations seem misplaced.

/lgtm

Thanks also for the code changes!

**Comment by justinsb** [comment]
@k8s-bot pull-kubernetes-e2e-kops-aws test this

**Comment by emmanuel** [comment]
Is this on track to land in 1.7.0? How about a backport to 1.6.x?

**Comment by emmanuel** [comment]
Also, while I'm asking questions... how does this work (w/r/t kube-proxy)? Are NodePort services' iptables rules set up to forward to a host-local pod? If not, how does this eliminate the extra hop?

**Comment by foolusion** [comment]
@emmanuel The beta local only annotation on the service or feature external traffic local only https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/iptables/proxier.go#L193

**Comment by MrHohn** [comment]
> Is this on track to land in 1.7.0? How about a backport to 1.6.x?

This is on track for 1.7.0 (code freeze will be on Thursday).

> Also, while I'm asking questions... how does this work (w/r/t kube-proxy)? Are NodePort services' iptables rules set up to forward to a host-local pod? If not, how does this eliminate the extra hop?

Exactly, kube-proxy sets up rules to forward relevant traffic to host-local pods only. More details could be found on [this proposal](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/external-lb-source-ip-preservation.md).

**Comment by emmanuel** [comment]
With the recent announcement of “Network Load Balancers” on AWS (L4 with source IP/port preservation: https://aws.amazon.com/blogs/aws/new-network-load-balancer-effortless-scaling-to-millions-of-requests-per-second/), true ESIPP is in reach on AWS. 

In other news, I believe that this PR can be closed, as this code shipped in 1.7.0.

**Pull request merged**